### ISSUE TITLE
xgboost: make sure thatr early stopping is properly supported by using a "watchtlist"

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr#1398]

**Comment by srijoyray** [comment]
+1
This feature is the only reason I keep going back to using xgboost package. 

Is there any hack or work around with the current way mlr is written to be able to pass the validation set to the watchlist parameter of xgboost?

**Comment by Stevo15025** [comment]
I took a stab at solving this within `trainLearner.regr.xgboost()`, though I wonder if that's the appropriate place. Sadly I'm getting the error

```r
# Error in check.custom.obj() : 
#   Setting objectives in 'params' and 'obj' at the same time is not allowed
```

Code:

I make a [0,1] bounded numeric learner in the `makeRLearner`  and use that for taking watchlist * length(target) samples for the watchlist. It's rather crude, but I think a better method would require something being done in the task.

```r
...
makeNumericLearnerParam(id = "watchlist", lower = 0, upper = 1),
...
```
Code:

```r
#' @export
trainLearner.regr.xgboost = function(.learner, .task, .subset, .weights = NULL,  ...) {
  parlist = list(...)

  parlist$label = getTaskData(.task, .subset, target.extra = TRUE)$target
  parlist$data = data.matrix(getTaskData(.task, .subset, target.extra = TRUE)$data)

  if (!is.null(parlist$watchlist)){
    watch_func <- function(x, y , watch.list){
      sampling = sample(length(y), floor(watch.list * length(y)))
      list(watch.list.x = x[sampling,], watch.list.y = y[sampling],
           x = x[-sampling,], y = y[-sampling], samples = sampling)
    }
    watch.list <- watch_func(x=parlist$data, y=parlist$label,
                             watch.list = parlist$watchlist)

    parlist$data = watch.list$x
    parlist$label = watch.list$y

    if (!is.null(.weights)){
      parlist$watchlist = list(validation1 = xgboost::xgb.DMatrix(data = watch.list$watch.list.x,
                                                                  label = watch.list$watch.list.y,
                                                                  weight = .weights[watch.list$samples]))
      .weights = .weights[-watch.list$samples]
    } else {
      parlist$watchlist = list(validation1 = xgboost::xgb.DMatrix(data = watch.list$watch.list.x,
                                                                  label = watch.list$watch.list.y))
    }
  }
  if (is.null(parlist$objective))
    parlist$objective = "reg:linear"

  if (!is.null(.weights))
    parlist$data = xgboost::xgb.DMatrix(data = parlist$data, label = parlist$label, weight = .weights)

  do.call(xgboost::xgboost, parlist)
}
```

**Comment by Stevo15025** [comment]
I think the best resolve for allowing early stopping is to have a separate model for tuning xgboost models through `xgb.cv()`

 `xgb.cv()` only asks for `folds` and `early_stopping_rounds` and then stops early based on the CV. So this removes the decision from us as to 'how and how much' to sample for the watchlist.

**Comment by chaoqing** [comment]
Have this feature been implemented? I actually mean, can we send the validation datasets into the train method when doing cross validation? For algorithms support early stop like xgboost, lightgbm, we need  to monitor its performance on validation datasets when `early_stop_round` is set. I know we can do this for a single training, but how do we do this when we are tuning the hyper parameter with `mlrMBO`?

**Comment by ja-thomas** [comment]
I have a version of xgboost with early stopping here: 
https://github.com/ja-thomas/autoxgboost/blob/master/R/RLearner_classif_xgboost.earlystop.R
https://github.com/ja-thomas/autoxgboost/blob/master/R/RLearner_regr_xgboost.earlystop.R

But this only supports passing an additional task as early stopping data.

These changes should go into mlr itself at some point. Handling early stopping consistently in mlr is something that would be great to have, so if anyone has spare time :)

**Comment by lorenzwalthert** [comment]
Until this gets fixed, I think at least a warning should be shown instead of "failing" silently. For me, this was completely unexpected and can have detrimental effects. Anyways, thanks to you guys for your ongoing work on the package.

**Comment by pat-s** [comment]
Such a feature should to into {mlr3}.
Since there seems to be quite a high demand for this, I'll open an issue in {mlr3leaners} but we probably won't add this feature to {mlr} anymore (I assume it would require quite some time).
